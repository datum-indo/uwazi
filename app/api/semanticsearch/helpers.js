import { search } from '../search';
import model from './model';
import resultsModel from './resultsModel';
import { COMPLETED } from './statuses';
import templates from '../templates';

export const getSearchDocuments = async ({ documents, query }, language, user) => {
  if (documents && documents.length) {
    return documents;
  }
  const _query = { ...query, limit: 9999, searchTerm: '' };
  const res = await search.search(_query, language, user);
  return res.rows.map(doc => doc.sharedId);
};

export const removePageAnnotations = text => text.replace(/\[\[\d+\]\]/g, '');

export const updateSearchDocumentStatus = async (searchId, sharedId, status) =>
  model.db.findOneAndUpdate(
    {
      _id: searchId,
      'documents.sharedId': sharedId,
    },
    {
      $set: { 'documents.$.status': status },
    },
    { new: true, lean: true }
  );

export const setSearchDocumentResults = async (searchId, sharedId, results) => {
  const averageScore = results.length
    ? results.reduce((total, curr) => total + curr.score, 0) / results.length
    : 0;
  const _results = [...results];
  const docResults = await resultsModel.db.findOneAndUpdate(
    {
      sharedId,
      searchId,
    },
    {
      sharedId,
      searchId,
      averageScore,
      results: _results.sort((r1, r2) => r2.score - r1.score),
      status: COMPLETED,
    },
    { upsert: true, new: true }
  );
  return docResults;
};

export const extractDocumentContent = async doc => {
  const { fullText, metadata } = doc;
  const contents = {};

  if (metadata) {
    const template = await templates.getById(doc.template);
    const metadataFields = template.properties.filter(prop => prop.type === 'markdown');

    metadataFields.forEach(field => {
      if (metadata[field.name]) {
        contents[field.name] = metadata[field.name];
      }
    });
  }

  if (fullText) {
    Object.keys(fullText).forEach(page => {
      contents[page] = removePageAnnotations(fullText[page]);
    });
  }

  return contents;
};
