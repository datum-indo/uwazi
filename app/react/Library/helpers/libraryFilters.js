import comonProperties from 'shared/comonProperties';
import prioritySortingCriteria from 'app/utils/prioritySortingCriteria';

function getOptions(property, thesauris) {
  const matchingTHesauri = thesauris.find(thesauri => thesauri._id === property.content);
  return matchingTHesauri ? matchingTHesauri.values : null;
}

export function populateOptions(filters, thesauris) {
  return filters.map(property => {
    if (property.content) {
      return Object.assign(property, { options: getOptions(property, thesauris) });
    }

    if (!property.content && property.type === 'relationship') {
      return Object.assign(property, {
        options: thesauris
          .filter(t => t.type === 'template')
          .reduce((options, thesauri) => options.concat(thesauri.values), []),
      });
    }

    if (!property.content && property.type === 'relationshipfilter') {
      property.filters = populateOptions(property.filters, thesauris);
    }

    return property;
  });
}

function URLQueryToState(query, templates, thesauris, relationTypes) {
  let properties = comonProperties.comonFilters(templates, relationTypes, query.types);

  if (!query.types || !query.types.length) {
    properties = comonProperties.defaultFilters(templates);
  }

  const {
    searchTerm = '',
    filters = {},
    sort = prioritySortingCriteria.get().sort,
    order = prioritySortingCriteria.get().order,
    userSelectedSorting,
  } = query;
  properties = populateOptions(properties, thesauris).map(property => {
    let defaultValue = {};

    if (property.type === 'text' || property.type === 'markdown') {
      defaultValue = '';
    }

    filters[property.name] = filters[property.name] ? filters[property.name] : defaultValue;
    return property;
  });
  return { properties, search: { searchTerm, filters, order, sort, userSelectedSorting } };
}

const getOptionCount = (aggregations, optionId, name) => {
  let aggregation;
  if (aggregations.all && aggregations.all[name]) {
    aggregation = aggregations.all[name].buckets.find(
      bucket => bucket.key.toString() === optionId.toString()
    );
  }
  return aggregation ? aggregation.filtered.doc_count : 0;
};

export function parseWithAggregations(filters, aggregations, showNoValue = true) {
  return filters.map(_property => {
    const property = Object.assign({}, _property);
    if (property.options && property.options.length) {
      if (showNoValue) {
        property.options.push({ id: 'missing', label: 'No Value', noValueKey: true });
      }
      property.options = property.options
        .map(_option => {
          const option = Object.assign(_option, {
            results: getOptionCount(aggregations, _option.id, property.name),
          });
          if (option.values) {
            option.values = option.values.map(_opt => {
              _opt.results = getOptionCount(aggregations, _opt.id, property.name);
              return _opt;
            });
          }
          return option;
        })
        .filter(opt => opt.results);
    }
    if (property.filters) {
      property.filters = parseWithAggregations(property.filters, aggregations, showNoValue);
    }
    return property;
  });
}

export default {
  URLQueryToState,
  populateOptions,
  parseWithAggregations,
};
