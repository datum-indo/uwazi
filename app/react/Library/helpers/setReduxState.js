import { wrapDispatch } from 'app/Multireducer';
import {
  setDocuments,
  unsetDocuments,
  initializeFiltersForm,
} from 'app/Library/actions/libraryActions';
import { actions as formActions } from 'react-redux-form';
import { actions } from 'app/BasicReducer';

export default function setReduxState(state) {
  return _dispatch => {
    const dispatch = wrapDispatch(_dispatch, 'library');
    dispatch(formActions.load('library.search', state.library.search));
    dispatch(unsetDocuments());

    dispatch(
      initializeFiltersForm({
        documentTypes: state.library.filters.documentTypes,
        libraryFilters: state.library.filters.properties,
        aggregations: state.library.aggregations,
      })
    );

    dispatch(setDocuments(state.library.documents));
    dispatch(actions.set('library.markers', state.library.markers));
  };
}
