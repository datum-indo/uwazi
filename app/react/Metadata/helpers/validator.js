export function notEmpty(val) {
  if (Array.isArray(val)) {
    return Boolean(val.length);
  }
  if (typeof val === 'number') {
    return true;
  }

  if (typeof val === 'object' && val !== null) {
    return Boolean(Object.keys(val).length);
  }
  return !!val && val.trim() !== '';
}

export default {
  generate(template, noTitle = false) {
    const validationObject = {
      title: { required: notEmpty },
    };

    if (noTitle) {
      delete validationObject.title;
    }

    template.properties.forEach(property => {
      if (property.required) {
        validationObject[`metadata.${property.name}`] = { required: notEmpty };
      }
    });

    return validationObject;
  },
};
