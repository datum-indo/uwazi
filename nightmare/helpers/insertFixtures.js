import { spawn } from 'child-process-promise';

export default async () => {
  try {
    const fixtures = await spawn('yarn', ['e2e-restore-fixtures'], {
      capture: ['stdout', 'stderr'],
    });
    console.log(fixtures.stdout);
  } catch (e) {
    process.exit(1);
  }
};
