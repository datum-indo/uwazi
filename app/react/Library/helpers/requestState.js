import prioritySortingCriteria from 'app/utils/prioritySortingCriteria';
import rison from 'rison';
import api from 'app/Search/SearchAPI';
import libraryHelpers from 'app/Library/helpers/libraryFilters';
import setReduxState from './setReduxState.js';

export function processQuery(_query, globalResources, key) {
  const defaultSearch = prioritySortingCriteria.get({ templates: globalResources.templates });

  let query;
  try {
    query = rison.decode(_query.q || '()');
  } catch (error) {
    error.status = 404;
    throw error;
  }

  query.order = query.order || defaultSearch.order;
  query.sort = query.sort || defaultSearch.sort;
  query.view = _query.view;

  if (key === 'markers') {
    query.geolocation = true;
  }
  const { userSelectedSorting, ...sanitizedQuery } = query;
  return sanitizedQuery;
}

export default function requestState(request, globalResources) {
  const documentsRequest = {
    ...request,
    data: processQuery(request.data, globalResources),
  };

  const markersRequest = {
    ...request,
    data: processQuery(request.data, globalResources, 'markers'),
  };

  return Promise.all([api.search(documentsRequest), api.search(markersRequest)]).then(
    ([documents, markers]) => {
      const filterState = libraryHelpers.URLQueryToState(
        documentsRequest.data,
        globalResources.templates.toJS(),
        globalResources.thesauris.toJS(),
        globalResources.relationTypes.toJS()
      );
      const state = {
        library: {
          filters: {
            documentTypes: documentsRequest.data.types || [],
            properties: filterState.properties,
          },
          aggregations: documents.aggregations,
          search: filterState.search,
          documents,
          markers,
        },
      };

      return [setReduxState(state)];
    }
  );
}
