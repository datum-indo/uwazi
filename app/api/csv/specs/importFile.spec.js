/** @format */

import path from 'path';
import fs from 'fs';
import importFile from '../importFile';
import { createTestingZip } from './helpers';

describe('importFile', () => {
  beforeAll(async () => {
    await createTestingZip(
      [path.join(__dirname, '/zipData/import.csv'), path.join(__dirname, '/zipData/file1.txt')],
      'ImportFile.zip'
    );

    await createTestingZip([path.join(__dirname, '/zipData/file1.txt')], 'badFile.zip');
  });

  afterAll(() => {
    fs.unlinkSync(path.join(__dirname, '/zipData/ImportFile.zip'));
    fs.unlinkSync(path.join(__dirname, '/zipData/badFile.zip'));
  });

  describe('readStream', () => {
    it('should return a readable stream for the csv file', async done => {
      const file = importFile(path.join(__dirname, '/test.csv'));
      let fileContents;
      (await file.readStream())
        .on('data', chunk => {
          fileContents += chunk;
        })
        .on('end', () => {
          expect(fileContents).toMatchSnapshot();
          done();
        });
    });

    describe('when file is a a zip', () => {
      it('should return a stream for the first csv file it encounters', async done => {
        const file = importFile(path.join(__dirname, '/zipData/ImportFile.zip'));
        let fileContents;
        (await file.readStream())
          .on('data', chunk => {
            fileContents += chunk;
          })
          .on('end', () => {
            expect(fileContents).toMatchSnapshot();
            done();
          });
      });

      describe('when passing a filename', () => {
        it('should return a read stream for that file', async done => {
          const file = importFile(path.join(__dirname, '/zipData/ImportFile.zip'));
          let fileContents;
          (await file.readStream('file1.txt'))
            .on('data', chunk => {
              fileContents += chunk;
            })
            .on('end', () => {
              expect(fileContents).toMatchSnapshot();
              done();
            });
        });
      });
    });
  });

  describe('when there is no import.csv on the zip', () => {
    it('should throw an error', async () => {
      const file = importFile(path.join(__dirname, '/zipData/badFile.zip'));
      let error;
      try {
        await file.readStream();
      } catch (e) {
        error = e;
      }

      expect(error.message.match(/import\.csv/)).not.toBeNull();
    });
  });
});
