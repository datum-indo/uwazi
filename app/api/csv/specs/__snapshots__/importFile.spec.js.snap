// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importFile readStream should return a readable stream for the csv file 1`] = `
"undefinedTitle , text label  , numeric label, non configured, select label, not defined type,

title1, text value 1, 1977, ______________, thesauri1   , notType1        ,
title2, text value 2, 2019, ______________, thesauri2   , notType2        ,
title3, text value 3, 2020, ______________, thesauri2   , notType3        ,
"
`;

exports[`importFile readStream when file is a a zip should return a stream for the first csv file it encounters 1`] = `
"undefinedTitle , file ,

file1, 1.pdf,
file2, 2.pdf,
file3, 3.pdf,
"
`;

exports[`importFile readStream when file is a a zip when passing a filename should return a read stream for that file 1`] = `
"undefinedfile1.txt
"
`;
